File: crates/yryvu-bridge/src/commands/conflicts.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::PathBuf;

use crate::backend::GitBackend;
use crate::repo::conflicts::{ConflictDiff3, ConflictListing, ConflictSide, ConflictSource};
use crate::repo::GixBackend;

#[tauri::command]
pub async fn list_conflicts(repo_path: String) -> Result<ConflictListing, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .list_conflicts(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn read_conflict_diff3(repo_path: String, path: String) -> Result<ConflictDiff3, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .read_conflict_diff3(&PathBuf::from(&repo_path), &path)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn accept_conflict_side(
    repo_path: String,
    path: String,
    side: ConflictSide,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .accept_conflict_side(&PathBuf::from(&repo_path), &path, side)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn resolve_conflict_with_content(
    repo_path: String,
    path: String,
    content: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .resolve_conflict_with_content(&PathBuf::from(&repo_path), &path, &content)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn mark_conflict_resolved(repo_path: String, path: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .mark_conflict_resolved(&PathBuf::from(&repo_path), &path)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn finish_in_progress_op(repo_path: String) -> Result<ConflictSource, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .finish_in_progress_op(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

File: apps/yryvu-app/src/ipc/conflicts.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

export type ConflictSource =
  | "merge"
  | "rebase"
  | "interactive-rebase"
  | "cherry-pick"
  | "revert"
  | "bisect"
  | "standalone";

export type ConflictSide = "ours" | "theirs" | "base";

export interface ConflictedFile {
  path: string;
  has_ancestor: boolean;
  has_ours: boolean;
  has_theirs: boolean;
}

export interface ConflictListing {
  source: ConflictSource;
  files: ConflictedFile[];
}

export interface ConflictDiff3 {
  base: string | null;
  ours: string | null;
  theirs: string | null;
  working: string;
}

export function listConflicts(repoPath: string) {
  return invoke<ConflictListing>("list_conflicts", { repoPath });
}

export function readConflictDiff3(repoPath: string, path: string) {
  return invoke<ConflictDiff3>("read_conflict_diff3", { repoPath, path });
}

export function acceptConflictSide(repoPath: string, path: string, side: ConflictSide) {
  return invoke<void>("accept_conflict_side", { repoPath, path, side });
}

export function resolveConflictWithContent(repoPath: string, path: string, content: string) {
  return invoke<void>("resolve_conflict_with_content", { repoPath, path, content });
}

export function markConflictResolved(repoPath: string, path: string) {
  return invoke<void>("mark_conflict_resolved", { repoPath, path });
}

export function finishInProgressOp(repoPath: string) {
  return invoke<ConflictSource>("finish_in_progress_op", { repoPath });
}
